src/app/missing_subchain/block.ml

```diff
@@ -0,0 +1,253 @@
+  ; staking_epoch_data=
+      mk_epoch_data block.staking_epoch_seed block.staking_epoch_ledger_hash
+  ; next_epoch_data=
+      mk_epoch_data block.next_epoch_seed block.next_epoch_ledger_hash
+  ; block_stake_winner= block.block_winner
+  ; block_creator= block.creator }
+
+let protocol_state_body_of_extensional_block (block : Extensional.Block.t) =
+  let blockchain_state = blockchain_state_of_extensional_block block in
+  let consensus_state = consensus_state_of_extensional_block block in
+  {blockchain_state; consensus_state}
+
+let protocol_state_of_extensional_block block =
+  {body= protocol_state_body_of_extensional_block block}
+
+let common_of_user_cmd (user_cmd : Extensional.User_command.t) =
+  { fee= user_cmd.fee
+  ; fee_token= user_cmd.fee_token
+  ; fee_payer_pk= user_cmd.fee_payer
+  ; nonce= user_cmd.nonce
+  ; valid_until= user_cmd.valid_until
+  ; memo= user_cmd.memo }
+
+let body_of_user_cmd (user_cmd : Extensional.User_command.t) =
+  let payload =
+    { source_pk= user_cmd.source
+    ; receiver_pk= user_cmd.receiver
+    ; token_id= user_cmd.token
+    ; amount= user_cmd.amount }
+  in
+  match user_cmd.typ with
+  | "payment" ->
+      Payment payload
+  | "delegation" ->
+      Stake_delegation payload
+  | "create_token" ->
+      Create_new_token payload
+  | "create_account" ->
+      Create_token_account payload
+  | "mint_tokens" ->
+      Mint_tokens payload
+  | cmd ->
+      failwithf "Unknown user command \"%s\"" cmd ()
+
+let payload_of_user_cmd user_cmd =
+  {common= common_of_user_cmd user_cmd; body= body_of_user_cmd user_cmd}
+
+let status_of_user_cmd (user_cmd : Extensional.User_command.t) =
+  let open Transaction_status in
+  (* balance data here is always empty, it's a dummy value
+     that data doesn't appear in archive db
+     this way, we can use the existing type Transaction_status.t
+  *)
+  match user_cmd.status with
+  | None ->
+      None
+  | Some "applied" ->
+      Some
+        (Applied
+           ( { fee_payer_account_creation_fee_paid=
+                 user_cmd.fee_payer_account_creation_fee_paid
+             ; receiver_account_creation_fee_paid=
+                 user_cmd.receiver_account_creation_fee_paid
+             ; created_token= user_cmd.created_token }
+           , Balance_data.empty ))
+  | Some "failed" -> (
+    match user_cmd.failure_reason with
+    | Some reason ->
+        Some (Failed (reason, Balance_data.empty))
+    | None ->
+        failwith "Failed user command status, no reason given" )
+  | Some s ->
+      failwithf "Unexpected user command status \"%s\"" s ()
+
+let signed_cmd_with_status_of_user_cmd (user_cmd : Extensional.User_command.t)
+    =
+  { With_opt_status.data= Signed_command {payload= payload_of_user_cmd user_cmd}
+  ; status= status_of_user_cmd user_cmd }
+
+let user_commands_of_extensional_block (block : Extensional.Block.t) =
+  let user_cmds = State_hash.Table.find_exn user_cmds_tbl block.state_hash in
```

Comment:
Can `user_cmds_tbl` be a parameter to this function instead of a global mutable?